mitodl/mit-open
# Generate OpenAPI spec, Type definitions, and API Client

#### What are the relevant tickets?
#5, See also https://github.com/mitodl/hq/discussions/2076

#### What's this PR do?
With this PR, we:
- commit an OpenAPI specification file (`openapi.yaml`) generated by [drf-spectacular](https://drf-spectacular.readthedocs.io/en/latest/).
- Use `openapi.yaml` to generate, via [OpenAPITools/openapi-generator](https://github.com/OpenAPITools/openapi-generator) (also committed in the repo):
    - Typescript type declarations for the API request/response objects
    - Runtime Typescript code for an API client using `axios`.
- Add CI checks to ensure the auto-generated, committed code is up-to-date.

**Note:** Currently the generated code is based on `channels_fields.urls` only. This is irrelevant and will change after #25 is merged.
 
#### How should this be manually tested?
On this branch:
1. Run `./scripts/generate_openapi.sh`. There should be no change.
2. Make an edit to `channel_fields/serializers.py`, for example:
    - change a serializer docstring
    - change the type annotation on a `SerializerMethodField`, e.g., make `get_full_name` return an `int` (which is wrong, but that's fine for now). 
3. **Check re-generating the code.** Re-run `./scripts/generate_openapi.sh` and inspect the git diff. You should see:
    - `openapi.yaml` has changed
    -  `/frontends/api/src/generated` has changed
4. Now check CI. We want to ensure CI fails if **either** `openapi.yaml` **or** `frontends/api/src/generated` is out of date.
5. (CI check):
    - `git checkout -b your-new-branch`.
    - commit just the serializer change, and push up to github.
    - The Python tests should fail because Django & `openapi.yaml` are out of sync. (*The new `openapi-generated-client-check` test will pass, since openapi.yaml is in-sync with the generated typescript.*)
        - Check the test failure message is useful.
6. (CI check, continued):
    - commit the freshly generated `openapi.yaml` spec, and push it up.
    - The Python tests should pass (Django and OpenAPI spec are in sync).
    - The new `openapi-generated-client-check` test will fail fairly quickly.
        - Check the test failure message is useful.
